src/v/redpanda/admin/kafka_connections_service.cc

```diff
@@ -96,6 +102,85 @@ ss::future<size_t> gather_all_shards(
 
 } // namespace
 
+kafka_connections_service::kafka_connections_service(
+  ss::sharded<kafka::server>& kafka_server)
+  : _kafka_server(kafka_server) {
+    if (ss::this_shard_id() == rate_limiter_shard) {
+        constexpr static auto max_requests_in_flight = size_t{1};
+        _rate_limiter = std::make_unique<ssx::semaphore>(
+          max_requests_in_flight, "kafka_connections_service/rate_limiter");
+    }
+}
+
+ss::future<>
+kafka_connections_service::start(ssx::semaphore& admin_memory_semaphore) {
+    _admin_memory_semaphore = &admin_memory_semaphore;
+    _admin_memory_semaphore_max = admin_memory_semaphore.current();
+    co_return;
+}
+
+ss::future<kafka_connections_service::remote_units>
+kafka_connections_service::rate_limit() {
+    // TODO: this could be optimized later by reserving memory on-demand, while
+    // iterating through shards.
+    co_return co_await container().invoke_on(
+      rate_limiter_shard,
+      [](kafka_connections_service& self) { return self.do_rate_limit(); });
+}
+
+ss::future<kafka_connections_service::remote_units>
+kafka_connections_service::do_rate_limit() {
+    vassert(
+      ss::this_shard_id() == rate_limiter_shard,
+      "do_rate_limit() must be called on shard {}",
+      rate_limiter_shard);
+    const auto timeout = std::chrono::milliseconds(
+      random_generators::get_int(3000, 5000));
```

Comment:
nit: normally we use https://github.com/redpanda-data/redpanda/blob/dev/src/v/random/simple_time_jitter.h for this